### PR TITLE
Sovereign tx preprocessor

### PR DESCRIPTION
**Pull request by mariusmihaic**
## Reasoning behind the pull request
- Reverted back changes to sovereign tx pre processor from this PR: https://github.com/multiversx/mx-chain-go/pull/5236. Therefore, one **should not review** `process/block/preprocess/sovereignChainTransactions.go` and `process/block/preprocess/sovereignChainTransactions_test.go`, since those files have been already reviewed before. You can simply change that the file content is unaltered with a compare tool
  
## Proposed changes
- Injected factory for tx pre processor
- There is already a task for `process/coordinator/sovereignChainProcess.go` tests, so those will be added in another PR

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: process/block/sovereignChainBlock.go [hunk 1]
```diff
@@ -609,11 +609,9 @@ func (scbp *sovereignChainBlockProcessor) ProcessBlock(headerHandler data.Header
 		return nil, nil, err
 	}
 
-	for _, accounts := range scbp.accountsDB {
-		if accounts.JournalLen() != 0 {
-			log.Error("sovereignChainBlockProcessor.ProcessBlock first entry", "stack", accounts.GetStackDebugFirstEntry())
-			return nil, nil, process.ErrAccountStateDirty
-		}
+	if scbp.accountsDB[state.UserAccountsState].JournalLen() != 0 {
```
Comment (sasurobert) [inline process/block/sovereignChainBlock.go:612]: check both peer accounts and user account.

**Review by sasurobert — APPROVED** [review]

File: process/coordinator/sovereignChainProcess.go [hunk 2]
```diff
@@ -168,3 +168,42 @@ func (sctc *sovereignChainTransactionCoordinator) CreateMbsAndProcessCrossShardT
 
 	return createMBDestMeExecutionInfo.miniBlocks, createMBDestMeExecutionInfo.numTxAdded, allMBsProcessed, nil
 }
+
+// CreateMbsAndProcessTransactionsFromMe should only create miniblocks and not process transactions
+func (sctc *sovereignChainTransactionCoordinator) CreateMbsAndProcessTransactionsFromMe(
```
Comment (SebastianMarian) [inline process/coordinator/sovereignChainProcess.go:173]: I don't think that this overwritten method is really necessary, as the only difference between this one and the main implementation is the following code: 
```
	interMBs := tc.CreatePostProcessMiniBlocks()
	if len(interMBs) > 0 {
		miniBlocks = append(miniBlocks, interMBs...)
	}

```
Which should always return an empty slice, as there is no execution done by the proposer at the time of block creation. Anyway, should be tested, but I think it is better to extract only the above code in separated methods for main and for sovereign, and keep the most of the common code only in one place (mainchain).

File: process/factory/metachain/preProcessorsContainerFactory_test.go [hunk 3]
```diff
@@ -4,7 +4,9 @@ import (
 	"testing"
 
 	"github.com/multiversx/mx-chain-go/dataRetriever"
+	errorsMx "github.com/multiversx/mx-chain-go/errors"
```
Comment (SebastianMarian) [inline process/factory/metachain/preProcessorsContainerFactory_test.go:7]: This alias is really needed?

File: process/factory/metachain/preProcessorsContainerFactory_test.go [hunk 4]
```diff
@@ -252,6 +256,17 @@ func TestPreProcessorsContainerFactory_CreateErrTxExecutionOrderHandler(t *testi
 	assert.Nil(t, ppcf)
 }
 
+func TestPreProcessorsContainerFactory_NilErrNilTxPreProcessorCreator(t *testing.T) {
```
Comment (SebastianMarian) [inline process/factory/metachain/preProcessorsContainerFactory_test.go:259]: TestPreProcessorsContainerFactory_NilTxPreProcessorCreator

File: process/block/sovereignChainBlock.go (hunk 1 above)
Comment (mariusmihaic) [reply process/block/sovereignChainBlock.go:612]: Reverted changes to check for all accounts in accounts db

File: process/block/preprocess/interfaces.go [hunk 5]
```diff
@@ -55,3 +56,9 @@ type BalanceComputationHandler interface {
 	AddressHasEnoughBalance(address []byte, value *big.Int) bool
 	IsInterfaceNil() bool
 }
+
+// TxPreProcessorCreator defines a tx pre processor creator
+type TxPreProcessorCreator interface {
+	CreateTxProcessor(args ArgsTransactionPreProcessor) (process.PreProcessor, error)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/interfaces.go:62]: Right, changed

File: process/block/preprocess/sovereignTxPreProcessorCreator.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package preprocess
+
+import "github.com/multiversx/mx-chain-go/process"
+
+type sovereignTxPreProcessorCreator struct {
+}
+
+// NewSovereignTxPreProcessorCreator creates a sovereign tx processor creator
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/sovereignTxPreProcessorCreator.go:8]: Right, changed

File: process/block/preprocess/sovereignTxPreProcessorCreator.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+func NewSovereignTxPreProcessorCreator() *sovereignTxPreProcessorCreator {
+	return &sovereignTxPreProcessorCreator{}
+}
+
+// CreateTxProcessor creates a tx processor creator for sovereign chain
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/sovereignTxPreProcessorCreator.go:13]: Right, changed

File: process/block/preprocess/sovereignTxPreProcessorCreator_test.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+func TestNewSovereignTxPreProcessorCreator(t *testing.T) {
+	t.Parallel()
+
+	creator := NewSovereignTxPreProcessorCreator()
+	require.False(t, creator.IsInterfaceNil())
+	require.Implements(t, new(TxPreProcessorCreator), creator)
+}
+
+func TestSovereignTxPreProcessorCreator_CreateTxProcessor(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/sovereignTxPreProcessorCreator_test.go:17]: Changed

File: process/block/preprocess/sovereignTxPreProcessorCreator_test.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	t.Parallel()
+
+	creator := NewSovereignTxPreProcessorCreator()
+	args := createDefaultTransactionsProcessorArgs()
+	txProc, err := creator.CreateTxProcessor(args)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/sovereignTxPreProcessorCreator_test.go:22]: Changed

File: process/block/preprocess/txPreProcessorCreator.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package preprocess
+
+import "github.com/multiversx/mx-chain-go/process"
+
+type txPreProcessorCreator struct {
+}
+
+// NewTxPreProcessorCreator creates a tx processor creator
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/txPreProcessorCreator.go:8]: Changed

File: process/block/preprocess/txPreProcessorCreator.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+func NewTxPreProcessorCreator() *txPreProcessorCreator {
+	return &txPreProcessorCreator{}
+}
+
+// CreateTxProcessor creates a tx processor creator for regular chain(meta+shard)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/txPreProcessorCreator.go:13]: Changed

File: process/block/preprocess/txPreProcessorCreator_test.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+func TestNewTxPreProcessorCreator(t *testing.T) {
+	t.Parallel()
+
+	creator := NewTxPreProcessorCreator()
+	require.False(t, creator.IsInterfaceNil())
+	require.Implements(t, new(TxPreProcessorCreator), creator)
+}
+
+func TestTxPreProcessorCreator_CreateTxProcessor(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/txPreProcessorCreator_test.go:17]: Changed

File: process/block/preprocess/txPreProcessorCreator_test.go [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+	t.Parallel()
+
+	creator := NewTxPreProcessorCreator()
+	args := createDefaultTransactionsProcessorArgs()
+	txProc, err := creator.CreateTxProcessor(args)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/txPreProcessorCreator_test.go:22]: Changed

File: process/factory/metachain/preProcessorsContainerFactory_test.go (hunk 3 above)
Comment (mariusmihaic) [reply process/factory/metachain/preProcessorsContainerFactory_test.go:7]: Not needed, changed it

File: process/factory/metachain/preProcessorsContainerFactory_test.go (hunk 4 above)
Comment (mariusmihaic) [reply process/factory/metachain/preProcessorsContainerFactory_test.go:259]: Right, changed

File: process/coordinator/sovereignChainProcess.go (hunk 2 above)
Comment (mariusmihaic) [reply process/coordinator/sovereignChainProcess.go:173]: Right, removed it, tested it and it worked

**Review by SebastianMarian — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**